server/src/main/java/org/apache/druid/client/SegmentLoadInfo.java

```diff
@@ -59,6 +61,15 @@ public ImmutableSegmentLoadInfo toImmutableSegmentLoadInfo()
     return new ImmutableSegmentLoadInfo(segment, servers);
   }
 
+  /**
+   * Randomly return one server from the sets of {@code servers}
+   */
+  public DruidServerMetadata pickOne()
+  {
+    synchronized (this) {
```

Comment:
if all of the methods are synchronized, then we don't need a concurrent set backing the servers list, suggest either dropping the concurrent backing set or making random pick not require locks like this, i don't have a strong opinion either way

Reply:
> hmm, this still isn't quite right i think, 

yes, I must synchronise the other methods as well. 
Or, should we continue to use the concurrent set and synchronise only `add`, `remove` and `pickOne`? 

> So I think we can just iterate to pick one...

`Iterators.get(servers.iterator(), ThreadLocalRandom.current().nextInt(servers.size()));`

Why do you suggest that we don't need synchronisation for this method. 
The `servers` size can change while iteration, if the server is removed then it can potentially result in `IndexOutOfBoundsException`. 

> but if we do, then something should probably be pushing a computation into this class to do the thing so that we can do the whole operation under lock...

can you please clarify this statement?

Reply:
yea, something similar to what @cryptoe suggests; `Iterators.get` is just taking the iterator returned by the collection and calling hasNext to navigate to a specific index, so we could just return the last element available if the index we picked is no longer present (though a while loop might be a bit cleaner)